File: LeetCode/1206. Design Skiplist/solution.py
# class Skiplist:

#     def __init__(self):
#         # random.seed(a=1023)
#         self.height = 0
#         self.head = Node()
#         self.head.nxt[-1] = self.head.pre[-1] = self.head

#     def search(self, target: int) -> bool:
#         # self.show()
#         cur = self.search_(target)
#         if cur != self.head and cur.val == target:
#             return True
#         return False

#     def search_(self, target):
#         idx = self.height
#         cur = self.head
#         while idx >= 0:
#             while cur.nxt[idx] != self.head and cur.nxt[idx].val <= target:
#                 cur = cur.nxt[idx]
#             idx -= 1
#         return cur

#     def coinFlip(self):
#         prob = 0.5
#         return random.uniform(0, 1) < prob

#     # def show(self):
#     #     head = self.head
#     #     height = len(head.nxt)
#     #     for idx in range(height):
#     #         cur = head
#     #         level = [cur.val]
#     #         while cur.nxt[idx] != self.head:
#     #             tmp_ = cur.nxt[0]
#     #             while tmp_ != cur.nxt[idx]:
#     #                 level.append('#')
#     #                 tmp_ = tmp_.nxt[0]
#     #             level.append(cur.nxt[idx].val)
#     #             cur = cur.nxt[idx]
#     #         print('level', idx, level)
#     #     print('\n')
#     #     return

#     def add(self, num: int) -> None:
#         # self.show()
#         cur = self.search_(num)
#         node = Node(num)
#         self.insertAfter(cur, node)
#         while self.coinFlip():
#             node.nxt.append(None)
#             node.pre.append(None)
#             while cur != self.head and len(node.nxt) > len(cur.nxt):
#                 cur = cur.pre[-1]
#             if cur == self.head and len(node.nxt) > len(cur.nxt):
#                 cur.nxt.append(self.head)
#                 cur.pre.append(self.head)
#             self.insertAfter(cur, node)
#         self.height = len(self.head.nxt) - 1
#         return

#     def insertAfter(self, cur, node):
#         idx = len(node.nxt) - 1
#         pre, nxt = cur, cur.nxt[idx]
#         node.nxt[idx], node.pre[idx] = nxt, pre
#         nxt.pre[idx], pre.nxt[idx] = node, node
#         return

#     def erase(self, num: int) -> bool:
#         # self.show()
#         cur = self.search_(num)
#         if cur != self.head and cur.val == num:
#             self.delete(cur)
#             return True
#         return False

#     def delete(self, node):
#         h = len(node.nxt)
#         for idx in range(h):
#             pre, nxt = node.pre[idx], node.nxt[idx]
#             pre.nxt[idx], nxt.pre[idx] = nxt, pre
#         return
    
# class Node:
#     def __init__(self, val=None):
#         self.val = val
#         self.pre = [None]
#         self.nxt = [None]


# # Your Skiplist object will be instantiated and called as such:
# # obj = Skiplist()
# # param_1 = obj.search(target)
# # obj.add(num)
# # param_3 = obj.erase(num)


class Node:
    __slots__ = 'val', 'levels'
    def __init__(self, val, levels):
        self.val = val
        self.levels = [None] * levels

class Skiplist(object):
    def __init__(self):
        self.head = Node(-1, 16) 
    
    def _iter(self, num):
        cur = self.head
        for level in range(15, -1, -1):
            while True:
                future = cur.levels[level]
                if future and future.val < num:
                    cur = future
                else:
                    break
            yield cur, level

    def search(self, target):
        for prev, level in self._iter(target):
            pass
        cur = prev.levels[0]
        return cur and cur.val == target

    def add(self, num):
        nodelvls = min(16, 1 + int(math.log2(1.0 / random.random())))
        node = Node(num, nodelvls)
        
        for cur, level in self._iter(num):
            if level < nodelvls:
                future = cur.levels[level]
                cur.levels[level] = node
                node.levels[level] = future

    def erase(self, num):
        ans = False
        for cur, level in self._iter(num):
            future = cur.levels[level]
            if future and future.val == num:
                ans = True
                cur.levels[level] = future.levels[level]
        return ans

File: LeetCode/1206. Design Skiplist/solution.cpp
#include<iostream>
#include<string>
#include<set>
#include<time.h>
#include <limits.h>

using namespace std;
template<typename T>
struct Less {
    bool operator () (const T & a , const T & b) const {
        return a < b;
    }
};
template<typename K, typename V,typename Comp = Less<K>>
class skip_list {
protected:
    struct skip_list_node {
        int level;
        const K key;
        V value;
        skip_list_node** forward;
        skip_list_node() :key{ 0 }, value{ 0 }, level{ 0 }, forward{0} {}
        skip_list_node(K k, V v, int l, skip_list_node* nxt = nullptr) :key(k), value(v), level(l) {
            forward = new skip_list_node * [level + 1];
            for (int i = 0; i <= level; ++i) forward[i] = nxt;
        }
        ~skip_list_node() { delete[] forward; }
    };
    using node = skip_list_node;
    void init() {
        srand((uint32_t)time(NULL));
        level = length = 0;
        head->forward = new node * [MAXL + 1];
        for (int i = 0; i <= MAXL; i++)
            head->forward[i] = tail;
    }
    int randomLevel() {
        int lv = 1; while ((rand() & S) < PS) ++lv;
        return MAXL > lv ? lv : MAXL;
    }
    int level;
    int length;
    static const int MAXL = 32;
    static const int P = 4;
    static const int S = 0xFFFF;
    static const int PS = S / P;
    static const int INVALID = INT_MAX;
    node* head, * tail;
    Comp less;
    node* find(const K& key, node** update) {
        node* p = head;
        for (int i = level; i >= 0; i--) {
            while (p->forward[i] != tail && less(p->forward[i]->key, key)) {
                p = p->forward[i];
            }
            update[i] = p;
        }
        p = p->forward[0];
        return p;
    }
    
public:
    struct Iter {
        node* p;
        Iter() : p(nullptr) {};
        Iter(node* rhs) : p(rhs) {}
        node* operator ->()const { return (p);}
        node& operator *() const { return *p;}
        bool operator == (const Iter& rhs) { return rhs.p == p;}
        bool operator != (const Iter& rhs) {return !(rhs.p == p);}
        void operator ++() {p = p->forward[0];}
        void operator ++(int) { p = p->forward[0]; }
    };
   
    skip_list() : head(new node()), tail(new node()), less{Comp()} {
        init();    
    }
    skip_list(Comp _less) : head(new node()), tail(new node()),  less{_less} {
        init();
    }
    void debug() {
        for (int i = MAXL; i >= 0; i--) {
            node *p = head;
            while (p != tail)
            {
                cout << p->key << " ";
                p = p->forward[i];
            }
            cout << endl;
        }
    }

    void insert(const K& key, const V& value) {
        node * update[MAXL + 1];
        node* p = find(key,update);
        // if (p != tail && p->key == key) {
        //     p->value = value;
        //     return;
        // }
        int lv = randomLevel();
        if (lv > level) {
            lv = ++level;
            update[lv] = head;
        }
        node * newNode = new node(key, value, lv);
        for (int i = lv; i >= 0; --i) {
            p = update[i];
            newNode->forward[i] = p->forward[i];
            p->forward[i] = newNode;
        }
        ++length;
    }

    bool erase(const K& key) {
        node* update[MAXL + 1];
        node* p = find(key, update);
        if (p->key != key)return false;
        for (int i = 0; i <= p->level; ++i) {
            update[i]->forward[i] = p->forward[i];
        }
        delete p;
        while (level > 0 && head->forward[level] == tail) --level;
        --length;
        return true;
    }
    Iter find(const K&key) {
        node* update[MAXL + 1];
        node* p = find(key, update);
        if (p == tail)return tail;
        if (p->key != key)return tail;
        return Iter(p);
    }
    bool count(const K& key) {
        node* update[MAXL + 1];
        node* p = find(key, update);
        if (p == tail)return false;
        return key == p->key;
    }
    Iter end() {
        return Iter(tail);
    }   
    Iter begin() {
        return Iter(head->forward[0]);
    }
};

class Skiplist: public skip_list<int,int> {
public:
    Skiplist() : skip_list<int,int>() {

    }
    
    bool search(int target) {
        return count(target);
    }
    
    void add(int num) {
        insert(num,0);
    }
    
    bool erase(int num) {
        return skip_list<int,int>::erase(num);
    }

    void debug() {
        return skip_list<int,int>::debug();
    }
};



const int maxn = 20;
const int inf = 0x3f3f3f3f;

struct Node
{
    int val;
    vector<Node*> nxt;
    Node(int v)
        :val(v)
        {}
    Node(int v, int d)
        :val(v)
        {
            nxt.assign(d,nullptr);
        }
};


class Skiplist {
public:
    Node *head, *tail;
    Skiplist() {
        srand(time(NULL));
        head = new Node(-inf,maxn);
        tail = new Node(inf,maxn);
        for (int i = 0; i < maxn; i++)
            head->nxt[i] = tail;
    }

    bool flipcoin(){
        return (double)rand()/RAND_MAX > 0.5;
    }

    void iter(int target, vector<Node**> &pnxt){
        Node *cur = head;
        for (int i = maxn-1; i >= 0; i--){
            while (cur->nxt[i]->val < target)
                cur = cur->nxt[i];
            pnxt[i] = &(cur->nxt[i]);
        }
    }
    
    bool search(int target) {
        vector<Node**> pnxt(maxn);
        iter(target, pnxt);
        return (*pnxt[0])->val == target;
    }
    
    void add(int num) {
        vector<Node**> pnxt(maxn);
        iter(num, pnxt);
        Node *m = new Node(num);
        for (int i = 0; i < maxn; i++){
            if (i && !flipcoin())
                break;
            m->nxt.push_back(*pnxt[i]);
            *pnxt[i] = m;
        }
    }
    
    bool erase(int num) {
        vector<Node**> pnxt(maxn);
        iter(num, pnxt);
        Node *m = *pnxt[0];
        if (m->val != num)
            return false;
        for (int i = 0; i < maxn; i++){
            if (i == (m->nxt).size())
                break;
            *pnxt[i] = m->nxt[i];
        }
        delete m;
        return true;
    }
};

/**
 * Your Skiplist object will be instantiated and called as such:
 * Skiplist* obj = new Skiplist();
 * bool param_1 = obj->search(target);
 * obj->add(num);
 * bool param_3 = obj->erase(num);
 */




const int maxn = 15;
const int inf = 0x3f3f3f3f;

struct Node
{
    int val;
    Node* nxt[maxn];
    Node(int v)
        :val(v)
        {
            memset(nxt, 0, maxn*sizeof(Node*));
        }
};


class Skiplist {
public:
    Node *head, *tail;
    Skiplist() {
        srand(time(NULL));
        head = new Node(-inf);
        tail = new Node(inf);
        for (int i = 0; i < maxn; i++)
            head->nxt[i] = tail;
    }

    bool flipcoin(){
        return (double)rand()/RAND_MAX > 0.5;
    }

    void iter(int target, Node** pnxt[]){
        Node *cur = head;
        for (int i = maxn-1; i >= 0; i--){
            while (cur->nxt[i]->val < target)
                cur = cur->nxt[i];
            pnxt[i] = &(cur->nxt[i]);
        }
    }
    
    bool search(int target) {
        Node** pnxt[maxn];
        iter(target, pnxt);
        return (*pnxt[0])->val == target;
    }
    
    void add(int num) {
        Node** pnxt[maxn];
        iter(num, pnxt);
        Node *m = new Node(num);
        for (int i = 0; i < maxn; i++){
            if (i && !flipcoin())
                break;
            m->nxt[i] = *pnxt[i];
            *pnxt[i] = m;
        }
    }
    
    bool erase(int num) {
        Node** pnxt[maxn];
        iter(num, pnxt);
        Node *m = *pnxt[0];
        if (m->val != num)
            return false;
        for (int i = 0; i < maxn; i++){
            if (!(m->nxt[i]))
                break;
            *pnxt[i] = m->nxt[i];
        }
        delete m;
        return true;
    }
};

/**
 * Your Skiplist object will be instantiated and called as such:
 * Skiplist* obj = new Skiplist();
 * bool param_1 = obj->search(target);
 * obj->add(num);
 * bool param_3 = obj->erase(num);
 */
